### ISSUE TITLE
[ENH]: set_horizontalalignment("right") on Y axis labels when yaxis.ticks_right() is used.

### ISSUE THREAD

**Issue by cmaydew** [matplotlib/matplotlib#21566]
### Problem

I have a subplot where the y axis labels need to be on the right side, however doing so results in left alignment of the labels. Setting them to right aligned puts them on top of the axis (as they are right aligned from the left edge of where they were left aligned). Of course setting a pad is an option, but that isn't dynamic to the labels. If I make a dozen graphs, they have a different number of digits, so the pad would need to be dynamic to that (and without using a monospaced font and counting the number of digits). 

### Proposed solution

I have made a way to make the dynamic padding, but it would likely need refinement before implementing. [Here is](https://stackoverflow.com/questions/68984904/matplotlib-tick-label-anchor-right-align-tick-labels-on-right-side-axis-and/69123387#69123387) a stack overflow question where I posted my method for right aligning and padding based on the width of the widest label. Even though I do have this current method, I'm submitting as a feature request since I haven't contributed to matplotlib before, and to allow for another solution that may be more refined.

**Comment by jklymak** [comment]
Can you clarify the proposed api?  How do you propose to tell the axis to do this?

**Comment by cmaydew** [comment]
I don't know, as I haven't dealt with that before. I was thinking just some sort of if statement based on the tick labels being on the right side and the horizontal alignment being right. But I am quite new to this and haven't really looked at the matplotlib code. So I'll take a few days to learn it and come up with implementation.

**Comment by tacaswell** [comment]
For right-aligning left tick or left-aligning right ticks it is relatively easy because we know where to anchor the text (some fixed offset from the axes independent of how long the text is).  We have tools (like tight layout and contrained layout) that will move the axes around to make sure everything fits, but at the end of the day the text is still a fixed stand off (we call it `pad`) from the Axes.  

See https://stackoverflow.com/questions/15882249/matplotlib-aligning-y-ticks-to-the-left/15883858#15883858 for how to do left-aligned tick labels on the left.  Doing right-aligned on the right should be the same (modulo possibly needing to change a sign).

**Comment by tacaswell** [comment]
🐑 Sorry for not reading very carefully and missing the SO link in the OP.

**Comment by jklymak** [comment]
Yes it's _possible_ but a) it would be a breaking change to not make it togglable and b) it may have performance implications so we may not want to do it without a toggle.  So I suspect it would need an API switch rather than matplotlib trying to guess what is desired.  

As @tacaswell points out above, the same issue occurs with left-aligned on the left axes.  Its not very numerate to align the left-most side of numbers, so it's a bit of a funny ask.

**Comment by timhoffm** [comment]
Right-aligning right labels (or left aligning left labels) is more more complicated. The alignment of text is relative to an anchor point. For tick labels these lie on a line parallel but offsetted to the axis. If you want to right-align right labels the offset becomes dependend on the size of all other tick labels. We currently do not have a mechanism to handle that.

Some ideas on possible ways forward:

- One would have to decide where the API and responsibility for such an operation would lie: Possibly with the Axis, because the individual texts don't have the knowledge of each other. Or one would have to wait for a redesign of the tick system to have TextCollections as ticks (maybe following up on https://github.com/timhoffm/matplotlib/commits/tick-refactor). Either way, this is a larger endeavour.
- A less invasive solution would be to encapsulate this in a standalone function similar to https://matplotlib.org/devdocs/gallery/pyplots/align_ylabels.html#sphx-glr-gallery-pyplots-align-ylabels-py The disadvantage is that this would only calculate fixed offsets and may not be correct anymore when the tick labels change e.g. because of zooming.
- Semi-OT: It may be helpful to introduce a notion of box-alignment additional to the anchor-point alignment


https://matplotlib.org/devdocs/gallery/pyplots/align_ylabels.html#sphx-glr-gallery-pyplots-align-ylabels-py

**Comment by jklymak** [comment]
@timhoffm I agree with all that. and I think _if_ there is ever a tick refactor, this may be a design criteria.

However, before we do backflips to make this happen, I am genuinely curious if there are published examples of this type of alignment?  Plotting packages don't align their ticks this way because they are too lazy to write the code to do the other alignment - they do it because it is far more clear to have the tick label directly adjoin the tick.   Take a look at hand-drawn graphs from before plotting packages - they almost all use the typical justification.  

This seems like something only a very small percentage of people want or should be using, so I'm not sure matplotlib should put a lot of effort into making this happen automagically, and rather folks should use the workarounds if the _really_ want this for some reason.

**Comment by timhoffm** [comment]
> I am genuinely curious if there are published examples of this type of alignment?

Indeed, this alignment is quite rare, but you'll sometimes find it, mostly for categorical data.

e.g.
- https://www.google.de/imgres?imgurl=https%3A%2F%2Fwww.europarl.europa.eu%2Fresources%2Flibrary%2Fimages%2F20210903PHT11251%2F20210903PHT11251_original.jpg&imgrefurl=https%3A%2F%2Fwww.europarl.europa.eu%2Fnews%2Fen%2Fpress-room%2F20210910IPR11916%2Fparliament-paves-the-way-for-special-fund-to-mitigate-the-impact-of-brexit&tbnid=m1gArohbX8BNbM&vet=10CEgQMyjWAmoQCgwICRUAAAAAHQAAAAAQCQ..i&docid=rC0iULlW4q_UhM&w=1201&h=1744&q=horizontal%20bar%20distribution&hl=de-de&client=safari&ved=0CEgQMyjWAmoQCgwICRUAAAAAHQAAAAAQCQ

And a case with center-alignment:
- https://www.google.de/imgres?imgurl=https%3A%2F%2Fpbs.twimg.com%2Fmedia%2FE0BprcdVoAAhQtt.jpg&imgrefurl=https%3A%2F%2Ftwitter.com%2Fbencowling88%2Fstatus%2F1387250562228391937&tbnid=-SNn_1yrE77QyM&vet=10CMkBEDMoqgJqEAoMCAkVAAAAAB0AAAAAEAk..i&docid=uBOb6fxV9ugoRM&w=1152&h=864&itg=1&q=horizontal%20bar%20distribution&hl=de-de&client=safari&ved=0CMkBEDMoqgJqEAoMCAkVAAAAAB0AAAAAEAk

As a first step, adding an example how to do this should be enough.

**Comment by timhoffm** [comment]
We can also be better in documenting what alignment means: #21571.

**Comment by cmaydew** [comment]
@timhoffm and @jklymak Thanks a lot for the discussion. That explains how difficult it could be to implement it without defining a separate function. As to the standalone function, if I'm not mistaken wouldn't that be a matter of when the function is called? Forgive me, as I'm quite new to coding, especially anything this complex, but is there any way to have the hypothetical function be run as part of the last step before plotting? I also agree that an example of how to do it is a great first step, I seem to recall stumbling on this when initially trying to do it and wishing there'd been an example as opposed to just the warning. So perhaps linking an example of how to do it here?
https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set_yticklabels.html
